assets/js/frontend/add-to-cart-variation.js

```diff
@@ -248,205 +228,265 @@
+						var variationAttributes = variations[ num ].attributes;
 
-						for ( var attr_name in attributes ) {
-							if ( attributes.hasOwnProperty( attr_name ) ) {
-								var attr_val = attributes[ attr_name ];
+						for ( var attr_name in variationAttributes ) {
+							if ( variationAttributes.hasOwnProperty( attr_name ) ) {
+								var attr_val         = variationAttributes[ attr_name ],
+									variation_active = '';
 
 								if ( attr_name === current_attr_name ) {
-
-									var variation_active = '';
-
 									if ( variations[ num ].variation_is_active ) {
 										variation_active = 'enabled';
 									}
 
 									if ( attr_val ) {
-
-										// Decode entities
+										// Decode entities and add slashes.
 										attr_val = $( '<div/>' ).html( attr_val ).text();
-
-										// Add slashes
 										attr_val = attr_val.replace( /'/g, '\\\'' );
 										attr_val = attr_val.replace( /"/g, '\\\"' );
 
-										// Compare the meerkat
-										current_attr_select.find( 'option[value="' + attr_val + '"]' ).addClass( 'attached ' + variation_active );
-
+										// Attach.
+										new_attr_select.find( 'option[value="' + attr_val + '"]' ).addClass( 'attached ' + variation_active );
 									} else {
-
-										current_attr_select.find( 'option' + option_gt_filter ).addClass( 'attached ' + variation_active );
-
+										// Attach all apart from placeholder.
+										new_attr_select.find( 'option:gt(0)' ).addClass( 'attached ' + variation_active );
 									}
 								}
 							}
 						}
 					}
 				}
 
-				// Detach unattached
-				current_attr_select.find( 'option' + option_gt_filter + ':not(.attached)' ).remove();
+				// Count available options.
+				attached_options_count = new_attr_select.find( 'option.attached' ).length;
+
+				// Check if current selection is in attached options.
+				if ( selected_attr_val && ( attached_options_count === 0 || new_attr_select.find( 'option.attached.enabled[value="' + selected_attr_val + '"]' ).length === 0 ) ) {
+					selected_attr_val_valid = false;
+				}
+
+				// Detach the placeholder if:
+				// - Valid options exist.
+				// - The current selection is non-empty.
+				// - The current selection is valid.
+				// - Placeholders are not set to be permanently visible.
+				if ( attached_options_count > 0 && selected_attr_val && selected_attr_val_valid && ( 'no' === show_option_none ) ) {
+					new_attr_select.find( 'option:first' ).remove();
+					option_gt_filter = '';
+				}
+
+				// Detach unattached.
+				new_attr_select.find( 'option' + option_gt_filter + ':not(.attached)' ).remove();
 
-				// Grey out disabled
-				current_attr_select.find( 'option' + option_gt_filter + ':not(.enabled)' ).attr( 'disabled', 'disabled' );
+				// Finally, copy to DOM and set value.
+				current_attr_select.html( new_attr_select.html() );
+				current_attr_select.find( 'option' + option_gt_filter + ':not(.enabled)' ).prop( 'disabled', true );
```

Comment:
Regarding `.attr` and `.prop`, I am a bit confused with which jQuery version deprecated `.attr`. My console is clean with this:

```
// Detach unattached.
new_attr_select.find( 'option' + option_gt_filter + ':not(.attached)' ).remove();

// Grey out disabled.
new_attr_select.find( 'option' + option_gt_filter + ':not(.enabled)' ).attr( 'disabled', 'disabled' );

// Finally, copy to DOM.
current_attr_select.html( new_attr_select.html() );
```

It would be nice if we could set the `disabled` attribute before updating the DOM.